File: backend/src/igab/api/v1/schemas/guide.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import Field, model_validator

from igab.api.v1.schemas.base import ApiModel


class ConceptInfo(ApiModel):
    """A concept the roadmap can ask about, and how it may be corrected."""

    key: str
    label: str
    kind: str
    binds_to: list[str]
    prompt: str
    caveat: str
    auto: bool
    allows_external: bool
    us_only: bool
    aliases: list[str]


class SignalResponse(ApiModel):
    key: str
    #: False when the user asked us to leave this concept alone.
    tracked: bool
    #: 'auto' | 'manual' | 'external' | 'manual+external' | 'dismissed'
    #: | 'answer' | 'off'
    source: str
    #: True/False when known, None when detection could not tell. Never guessed.
    met: bool | None = None
    #: Detected plus self-reported, in the concept's own units.
    value: Decimal | None = None
    detected_value: Decimal | None = None
    #: Self-reported and never blended into any report — Guide only.
    external_value: Decimal | None = None
    external_declared: bool = False
    external_as_of: date | None = None
    target: Decimal | None = None
    #: The emergency fund only: the starter cushion (the flat figure or one
    #: month of essentials, whichever is larger) and whether the fund clears
    #: it. The roadmap's starter step reads these; the full step reads
    #: `target` / `met`. None on every other concept.
    starter_target: Decimal | None = None
    starter_met: bool | None = None
    reason: str = ""
    entities: dict[str, list[str]] = Field(default_factory=dict)
    #: Things worth mentioning that did not count — a debt with no rate on
    #: record. A gap in the data is a nudge, not a silence.
    gaps: list[str] = Field(default_factory=list)
    note: str | None = None


class SignalsResponse(ApiModel):
    personalization: bool
    concepts: list[SignalResponse]


class CandidateOption(ApiModel):
    id: str
    name: str
    detail: str | None = None


class CandidatesResponse(ApiModel):
    concept_key: str
    options: dict[str, list[CandidateOption]]


class BindingUpdate(ApiModel):
    """One concept's whole answer — the endpoint replaces rather than merges."""

    #: 'auto' resets to the app's own guess and deletes every stored row.
    mode: str = Field(pattern="^(auto|manual|dismissed|answer|external)$")
    entity_ids: dict[str, list[uuid.UUID]] | None = None
    answer: bool | None = None
    #: Declare money held outside IGAB alongside anything bound above.
    external: bool = False
    #: Optional on purpose: "I have this covered" is a complete answer, and
    #: requiring a figure invites an invented one.
    external_amount: Decimal | None = Field(default=None, ge=0)
    note: str | None = Field(default=None, max_length=500)

    @model_validator(mode="after")
    def check_shape(self) -> "BindingUpdate":
        if self.mode == "answer" and self.answer is None:
            raise ValueError("answer is required when mode is 'answer'")
        if self.mode == "external" and not self.external:
            # Saying mode='external' without setting the flag would store
            # nothing and silently read as a reset.
            self.external = True
        return self


class PreferencesUpdate(ApiModel):
    personalization: bool | None = None
    checkup: bool | None = None
    wishlist: bool | None = None
    #: Consent to the money move turning the Wishlist off implies. Without it
    #: the server refuses while a wish envelope still holds a balance and says
    #: how much, so the dialog can state the figure and no request that merely
    #: says "wishlist: false" moves money on its own.
    release_wishlist_money: bool = False


class WishlistRetirePreview(ApiModel):
    """What turning the Wishlist off would return to Ready to Assign.

    Served rather than summed on the client from the month's balances: the
    switch and the endpoint must agree on the figure, and the client cannot see
    an archived group's envelopes to add them up in the first place.
    """

    #: Envelopes carrying a balance, named so the dialog can list them.
    envelopes: list[str]
    #: Their total, in the current month.
    available: Decimal
    #: Nothing to move — the switch may flip with no dialog at all.
    is_empty: bool


class PreferencesResponse(ApiModel):
    personalization: bool
    checkup: bool
    wishlist: bool


class StepUpdate(ApiModel):
    #: null clears the mark and returns the step to undecided.
    state: str | None = Field(default=None, pattern="^(done|skipped)$")


class GuideOverview(ApiModel):
    """Everything the Guide needs to render, in one round trip."""

    concepts: list[ConceptInfo]
    thresholds: dict[str, int]
    preferences: PreferencesResponse
    progress: dict[str, str]


class CheckupMetric(ApiModel):
    """One row of the checkup: a figure against the target the roadmap states."""

    key: str
    label: str
    value: Decimal | None = None
    target: Decimal | None = None
    #: 'money' | 'months' | 'percent' | 'count'
    unit: str
    detail: str = ""
    #: Finding kinds this row is the home of, so the client can mark it.
    finding_kinds: list[str] = Field(default_factory=list)
    #: A report tab with the numbers behind this row, when one exists.
    report: str | None = None
    #: What the figure counts, by name — the whole list; the client paces it.
    names: list[str] = Field(default_factory=list)
    #: The same figure in money when `unit` is not money — the emergency fund
    #: in months also says what those months are worth.
    money_value: Decimal | None = None
    money_target: Decimal | None = None


class CheckupFinding(ApiModel):
    kind: str
    rank: int
    concept_key: str | None = None
    title: str
    detail: str = ""
    value: Decimal | None = None
    target: Decimal | None = None
    names: list[str] = Field(default_factory=list)


class CheckupResponse(ApiModel):
    """Metrics, and every finding that fired, most severe first.

    All findings are returned: the roadmap's step markers need every kind, and
    how many the health report shows is the client's call.
    """

    enabled: bool
    as_of: date
    last_run: datetime | None = None
    metrics: list[CheckupMetric]
    findings: list[CheckupFinding]


# ── scenario calculators ─────────────────────────────────────────────────────
#
# Nothing here is persisted. Inputs are what the user typed (or what the
# planner seeded from their liabilities, which they may have edited); the
# server answers with arithmetic it can show its working for.

Money = Decimal


class CascadeDebtIn(ApiModel):
    key: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=120)
    balance: Money = Field(ge=0)
    annual_rate: Decimal = Field(ge=0, le=100)
    minimum_payment: Money = Field(ge=0)


class PayoffPlanRequest(ApiModel):
    debts: list[CascadeDebtIn] = Field(min_length=1, max_length=50)
    extra: Money = Field(default=Decimal("0"), ge=0)

    @model_validator(mode="after")
    def unique_keys(self) -> "PayoffPlanRequest":
        keys = [d.key for d in self.debts]
        if len(set(keys)) != len(keys):
            raise ValueError("debt keys must be unique")
        return self


class CascadeDebtOut(ApiModel):
    key: str
    name: str
    order: int
    payoff_date: date | None
    months: int
    never_pays_off: bool
    total_interest: Decimal
    total_principal: Decimal


class CascadeMonthOut(ApiModel):
    month_index: int
    date: date
    payment: Decimal
    principal_paid: Decimal
    interest_paid: Decimal
    balance: Decimal
    balances: dict[str, Decimal]


class CascadeOut(ApiModel):
    order: str
    debts: list[CascadeDebtOut]
    months: list[CascadeMonthOut]
    debt_free_date: date | None
    never_pays_off: bool
    total_interest: Decimal
    total_paid: Decimal


class PayoffPlanResponse(ApiModel):
    as_of: date
    extra: Decimal
    avalanche: CascadeOut
    snowball: CascadeOut
    #: Minimums only, nothing rolled — what happens if nothing changes.
    minimums_only: CascadeOut


class PayVsSaveRequest(ApiModel):
    balance: Money = Field(ge=0)
    annual_rate: Decimal = Field(ge=0, le=100)
    minimum_payment: Money = Field(ge=0)
    extra: Money = Field(ge=0)
    #: A rate the user can get today, typed by them. Never assumed.
    savings_apy: Decimal = Field(ge=0, le=100)


class PayVsSaveResponse(ApiModel):
    horizon_months: int
    baseline_total_interest: Decimal
    baseline_never_pays_off: bool
    pay_months: int
    pay_payoff_date: date | None
    pay_total_interest: Decimal
    pay_never_pays_off: bool
    debt_interest_saved: Decimal
    months_sooner: int
    savings_contributed: Decimal
    savings_balance: Decimal
    savings_interest_earned: Decimal
    breakeven_apy: Decimal | None
    favours: Literal["pay", "save", "even"]


class LoanIn(ApiModel):
    name: str = Field(min_length=1, max_length=120)
    principal: Money = Field(ge=0)
    annual_rate: Decimal = Field(ge=0, le=100)
    term_months: int | None = Field(default=None, ge=1, le=600)
    payment: Money | None = Field(default=None, ge=0)
    fees: Money = Field(default=Decimal("0"), ge=0)

    @model_validator(mode="after")
    def term_or_payment(self) -> "LoanIn":
        if self.term_months is None and self.payment is None:
            raise ValueError("give a term or a payment")
        return self


class LoanCompareRequest(ApiModel):
    loans: list[LoanIn] = Field(min_length=1, max_length=10)


class LoanOutcomeOut(ApiModel):
    name: str
    payment: Decimal
    months: int
    payoff_date: date | None
    never_pays_off: bool
    total_interest: Decimal
    total_cost: Decimal


class LoanCompareResponse(ApiModel):
    loans: list[LoanOutcomeOut]
    cheapest: str | None


class EmergencyFundRequest(ApiModel):
    months: int = Field(ge=1, le=12)
    monthly_contribution: Money = Field(default=Decimal("0"), ge=0)


class EmergencyFundResponse(ApiModel):
    """Sized from the roadmap's own essentials and emergency-fund figures."""

    months: int
    monthly_contribution: Decimal
    essentials_monthly: Decimal | None
    current: Decimal | None
    target: Decimal | None
    gap: Decimal | None
    months_to_fund: int | None
    funded_by: date | None

File: frontend/src/api/guide.ts
import { keepPreviousData, useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import toast from 'react-hot-toast'
import { apiClient, apiErrorMessage } from './client'
import type { SignalKey } from '../content/roadmap'
import { ROOT } from './queryKeys'

/** How a concept came to be answered. */
export type SignalSource =
  | 'auto'
  | 'manual'
  | 'external'
  | 'manual+external'
  | 'dismissed'
  | 'answer'
  /** Personalisation is switched off — no detection ran. */
  | 'off'

export type EntityType = 'category' | 'account' | 'liability'

export interface ConceptInfo {
  key: SignalKey
  label: string
  kind: 'amount' | 'rate' | 'boolean'
  binds_to: EntityType[]
  prompt: string
  caveat: string
  auto: boolean
  allows_external: boolean
  us_only: boolean
  aliases: string[]
}

export interface Signal {
  key: SignalKey
  tracked: boolean
  source: SignalSource
  /** null whenever detection could not tell. Never a guess. */
  met: boolean | null
  /** Detected plus self-reported. Strings: these are money, not floats. */
  value: string | null
  detected_value: string | null
  external_value: string | null
  external_declared: boolean
  external_as_of: string | null
  target: string | null
  /** Emergency fund only: the starter cushion ($1,000 or one month of
   *  essentials, whichever is larger) and whether the fund clears it. The
   *  roadmap's starter step reads these; the full step reads `target`/`met`. */
  starter_target: string | null
  starter_met: boolean | null
  reason: string
  entities: Partial<Record<EntityType, string[]>>
  /** Things that did not count and should be said out loud — a debt with no
   *  rate on record. A gap in the data is a nudge, not a silence. */
  gaps: string[]
  note: string | null
}

export interface SignalsResponse {
  personalization: boolean
  concepts: Signal[]
}

export interface GuidePreferences {
  personalization: boolean
  checkup: boolean
  wishlist: boolean
}

export interface GuideOverview {
  concepts: ConceptInfo[]
  thresholds: Record<string, number>
  preferences: GuidePreferences
  progress: Record<string, 'done' | 'skipped'>
}

export interface CandidateOption {
  id: string
  name: string
  detail?: string | null
}

export interface BindingUpdate {
  mode: 'auto' | 'manual' | 'dismissed' | 'answer' | 'external'
  entity_ids?: Partial<Record<EntityType, string[]>>
  answer?: boolean
  external?: boolean
  /** Optional on purpose — "I have this covered" is a complete answer. */
  external_amount?: string | null
  note?: string | null
}

export function useGuideOverview(budgetId: string | null) {
  return useQuery({
    queryKey: [ROOT.guide, budgetId],
    queryFn: () => apiClient.get<GuideOverview>(`/${budgetId}/guide`).then((r) => r.data),
    enabled: !!budgetId,
    staleTime: 60_000,
  })
}

export function useGuideSignals(budgetId: string | null, enabled = true) {
  return useQuery({
    queryKey: [ROOT.guideSignals, budgetId],
    queryFn: () =>
      apiClient.get<SignalsResponse>(`/${budgetId}/guide/signals`).then((r) => r.data),
    enabled: !!budgetId && enabled,
    // Signals are derived from the whole budget, so almost any edit could move
    // them. Short and refetched on demand rather than aggressively live.
    staleTime: 30_000,
  })
}

export function useConceptCandidates(budgetId: string | null, conceptKey: string | null) {
  return useQuery({
    queryKey: [ROOT.guideCandidates, budgetId, conceptKey],
    queryFn: () =>
      apiClient
        .get<{ concept_key: string; options: Partial<Record<EntityType, CandidateOption[]>> }>(
          `/${budgetId}/guide/candidates/${conceptKey}`
        )
        .then((r) => r.data.options),
    enabled: !!budgetId && !!conceptKey,
    staleTime: 60_000,
  })
}

export function useSetBinding(budgetId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ conceptKey, ...body }: BindingUpdate & { conceptKey: string }) =>
      apiClient.put(`/${budgetId}/guide/bindings/${conceptKey}`, body),
    onSuccess: () => qc.invalidateQueries({ queryKey: [ROOT.guideSignals, budgetId] }),
  })
}

/** What turning the Wishlist off would return to Ready to Assign.
 *
 *  Served, not summed here from the month's balances: the switch and the
 *  endpoint have to agree on the figure, and once the group is archived the
 *  client cannot see its envelopes to add them up at all. */
export interface WishlistRetirePreview {
  envelopes: string[]
  available: string
  is_empty: boolean
}

/** Asked at the moment the switch is clicked, not held in a cache: the answer
 *  is a money figure the dialog is about to state, and a stale one would name
 *  an amount that is no longer there. */
export function fetchWishlistRetirePreview(budgetId: string) {
  return apiClient
    .get<WishlistRetirePreview>(`/${budgetId}/guide/wishlist/retire-preview`)
    .then((r) => r.data)
}

export function useSetGuidePreferences(budgetId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (
      changes: Partial<GuidePreferences> & { release_wishlist_money?: boolean }
    ) =>
      apiClient
        .put<GuidePreferences>(`/${budgetId}/guide/preferences`, changes)
        .then((r) => r.data),
    onSuccess: (prefs) => {
      // Seed rather than only invalidate: the toggle should settle instantly,
      // and the server may have forced checkup off alongside personalisation.
      qc.setQueryData<GuideOverview | undefined>([ROOT.guide, budgetId], (old) =>
        old ? { ...old, preferences: prefs } : old
      )
      qc.invalidateQueries({ queryKey: [ROOT.guide, budgetId] })
      qc.invalidateQueries({ queryKey: [ROOT.guideSignals, budgetId] })
    },
    // The server refuses a wishlist-off that would move money without an
    // explicit confirmation, and its sentence carries the figure. A silent
    // failure here would look like a switch that simply does not work.
    onError: (e) => toast.error(apiErrorMessage(e, 'Could not save that setting')),
  })
}

export function useSetGuideStep(budgetId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ stageId, state }: { stageId: string; state: 'done' | 'skipped' | null }) =>
      apiClient.put(`/${budgetId}/guide/progress/${stageId}`, { state }),
    onSuccess: () => qc.invalidateQueries({ queryKey: [ROOT.guide, budgetId] }),
  })
}

// ── the checkup ──────────────────────────────────────────────────────────────

/** What the health report can find. Home of each kind → roadmap step:
 *  components/guide/checkupLeds.ts. */
export type FindingKind =
  | 'high_interest_debt'
  | 'ef_not_started'
  | 'ef_below_starter'
  | 'chronic_overspend'
  | 'ef_below_full'
  | 'moderate_debt'
  | 'retirement_below_target'
  | 'stale_external'
  | 'unknown_rates'

export interface CheckupFinding {
  kind: FindingKind
  /** Severity, 1 = most. The server sorts by it; the client shows the first few. */
  rank: number
  concept_key: string | null
  /** A short clause with no figures — compose with `value` in the budget's currency. */
  title: string
  detail: string
  value: string | null
  target: string | null
  names: string[]
}

export interface CheckupMetric {
  key: string
  label: string
  value: string | null
  target: string | null
  unit: 'money' | 'months' | 'percent' | 'count'
  detail: string
  /** Finding kinds this row is the home of — mark it when one fired. */
  finding_kinds: FindingKind[]
  /** A report tab with the numbers behind this row, when one exists. */
  report: string | null
  /** What the figure counts, by name — the whole list; NameChips paces it. */
  names: string[]
  /** The same figure in money when `unit` is not money — the emergency fund
   *  in months also says what those months are worth. */
  money_value: string | null
  money_target: string | null
}

export interface Checkup {
  enabled: boolean
  as_of: string
  last_run: string | null
  metrics: CheckupMetric[]
  /** Every finding that fired, most severe first — never capped by the server. */
  findings: CheckupFinding[]
}

export function useGuideCheckup(budgetId: string | null, enabled = true) {
  return useQuery({
    queryKey: [ROOT.guideCheckup, budgetId],
    queryFn: () => apiClient.get<Checkup>(`/${budgetId}/guide/checkup`).then((r) => r.data),
    // Gated on the preference by the caller: with reviews off, no request at all.
    enabled: !!budgetId && enabled,
    staleTime: 30_000,
  })
}

export function useRunHealthReport(budgetId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: () =>
      apiClient.post<Checkup>(`/${budgetId}/guide/checkup/run`).then((r) => r.data),
    // The run returns the same payload the GET would, freshly stamped.
    onSuccess: (checkup) => qc.setQueryData([ROOT.guideCheckup, budgetId], checkup),
  })
}


// ── scenario calculators ─────────────────────────────────────────────────────
// Inputs are documents the user typed (or edited from a seed), so these are
// POSTs behind useQuery: keyed on the body, previous answer kept while the next
// one is in flight. Money and rates travel as strings, never floats.

export interface CascadeDebtIn {
  key: string
  name: string
  balance: string
  annual_rate: string
  minimum_payment: string
}

export interface PayoffPlanRequest {
  debts: CascadeDebtIn[]
  extra: string
}

export interface CascadeDebtOut {
  key: string
  name: string
  order: number
  payoff_date: string | null
  months: number
  never_pays_off: boolean
  total_interest: string
  total_principal: string
}

export interface CascadeMonthOut {
  month_index: number
  date: string
  payment: string
  principal_paid: string
  interest_paid: string
  balance: string
  balances: Record<string, string>
}

export interface CascadeOut {
  order: 'avalanche' | 'snowball'
  debts: CascadeDebtOut[]
  months: CascadeMonthOut[]
  debt_free_date: string | null
  never_pays_off: boolean
  total_interest: string
  total_paid: string
}

export interface PayoffPlanResponse {
  as_of: string
  extra: string
  avalanche: CascadeOut
  snowball: CascadeOut
  /** Minimums only, nothing rolled — what happens if nothing changes. */
  minimums_only: CascadeOut
}

export interface PayVsSaveRequest {
  balance: string
  annual_rate: string
  minimum_payment: string
  extra: string
  savings_apy: string
}

export interface PayVsSaveResponse {
  horizon_months: number
  baseline_total_interest: string
  baseline_never_pays_off: boolean
  pay_months: number
  pay_payoff_date: string | null
  pay_total_interest: string
  pay_never_pays_off: boolean
  debt_interest_saved: string
  months_sooner: number
  savings_contributed: string
  savings_balance: string
  savings_interest_earned: string
  breakeven_apy: string | null
  favours: 'pay' | 'save' | 'even'
}

export interface LoanIn {
  name: string
  principal: string
  annual_rate: string
  term_months?: number | null
  payment?: string | null
  fees?: string
}

export interface LoanCompareRequest {
  loans: LoanIn[]
}

export interface LoanOutcomeOut {
  name: string
  payment: string
  months: number
  payoff_date: string | null
  never_pays_off: boolean
  total_interest: string
  total_cost: string
}

export interface LoanCompareResponse {
  loans: LoanOutcomeOut[]
  cheapest: string | null
}

export interface EmergencyFundRequest {
  months: number
  monthly_contribution: string
}

export interface EmergencyFundResponse {
  months: number
  monthly_contribution: string
  essentials_monthly: string | null
  current: string | null
  target: string | null
  gap: string | null
  months_to_fund: number | null
  funded_by: string | null
}

function useScenario<Req, Res>(kind: string, budgetId: string | null, body: Req | null) {
  return useQuery({
    queryKey: [ROOT.guideScenario, kind, budgetId, body],
    queryFn: () =>
      apiClient
        .post<Res>(`/${budgetId}/guide/scenarios/${kind}`, body)
        .then((r) => r.data),
    enabled: !!budgetId && body !== null,
    placeholderData: keepPreviousData,
    staleTime: 60_000,
  })
}

export function usePayoffPlan(budgetId: string | null, body: PayoffPlanRequest | null) {
  return useScenario<PayoffPlanRequest, PayoffPlanResponse>('payoff-plan', budgetId, body)
}

export function usePayVsSave(budgetId: string | null, body: PayVsSaveRequest | null) {
  return useScenario<PayVsSaveRequest, PayVsSaveResponse>('pay-vs-save', budgetId, body)
}

export function useLoanCompare(budgetId: string | null, body: LoanCompareRequest | null) {
  return useScenario<LoanCompareRequest, LoanCompareResponse>('loan-compare', budgetId, body)
}

export function useEmergencyFundPlan(
  budgetId: string | null,
  body: EmergencyFundRequest | null
) {
  return useScenario<EmergencyFundRequest, EmergencyFundResponse>('emergency-fund', budgetId, body)
}
